File: src/main/java/xyz/upperlevel/hgame/network/Endpoint.java
package xyz.upperlevel.hgame.network;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import io.netty.channel.Channel;
import io.netty.channel.EventLoopGroup;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import xyz.upperlevel.hgame.event.EventChannel;
import xyz.upperlevel.hgame.network.events.ConnectionCloseEvent;
import xyz.upperlevel.hgame.network.events.ConnectionOpenEvent;

public class Endpoint {
    @Getter
    private final Protocol protocol;
    @Getter
    private final ObjectMapper jsonMapper;
    @Getter
    private EventChannel events = new EventChannel();

    @Getter
    private Channel channel = null;

    @Setter(AccessLevel.PROTECTED)
    private EventLoopGroup eventGroup;

    public Endpoint(Protocol protocol) {
        this.protocol = protocol;
        jsonMapper = new ObjectMapper();

        SimpleModule module = new SimpleModule();
        module.addDeserializer(PayloadPacket.class, new PayloadPacketDeserializer(protocol));

        jsonMapper.registerModule(module);
    }


    public void onPacketReceive(PayloadPacket msg) {
        events.call(msg.getData());
    }

    public void onOpen(Channel channel) {
        this.channel = channel;
        events.call(new ConnectionOpenEvent());
    }

    public void onClose() {
        events.call(new ConnectionCloseEvent());
        this.channel = null;
    }

    public boolean isConnected() {
        return channel != null;
    }

    public void send(Packet packet) {
        var id = protocol.fromClass(packet.getClass())
                .orElseThrow(() -> new IllegalArgumentException("Packet not registered: " + packet.getClass()));
        channel.writeAndFlush(new PayloadPacket(id, packet));
    }

    public void close() throws InterruptedException {
        if (!isConnected()) return;
        getChannel().close().sync();
    }

    public void closeGracefully() throws InterruptedException {
        close();
        eventGroup.shutdownGracefully();
    }
}

File: src/main/java/xyz/upperlevel/hgame/network/Endpoint.kt
package xyz.upperlevel.hgame.network

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.module.SimpleModule
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import io.netty.channel.Channel
import io.netty.channel.EventLoopGroup
import xyz.upperlevel.hgame.event.EventChannel
import xyz.upperlevel.hgame.network.events.ConnectionCloseEvent
import xyz.upperlevel.hgame.network.events.ConnectionOpenEvent

abstract class Endpoint(val protocol: Protocol,
                        val side: NetSide) {
    val jsonMapper: ObjectMapper = jacksonObjectMapper()
    val events = EventChannel()

    var channel: Channel? = null
        private set

    protected var eventGroup: EventLoopGroup? = null

    val isConnected: Boolean
        get() = channel != null

    init {
        val module = SimpleModule()
        module.addDeserializer(PayloadPacket::class.java, PayloadPacketDeserializer(protocol))

        jsonMapper.registerModule(module)
    }


    open fun onPacketReceive(msg: PayloadPacket) {
        events.call(msg.data)
    }

    open fun onOpen(channel: Channel) {
        this.channel = channel
        events.call(ConnectionOpenEvent(channel))
    }

    open fun onClose() {
        events.call(ConnectionCloseEvent(channel!!))
        this.channel = null
    }

    open fun send(packet: Packet) {
        val id = protocol.fromClass(packet.javaClass)
                ?: throw IllegalArgumentException("Packet not registered: " + packet.javaClass)
        channel!!.writeAndFlush(PayloadPacket(id, packet))
    }

    abstract fun openAsync()

    open fun close() {
        if (!isConnected) return
        channel?.close()!!.sync()
    }

    fun closeGracefully() {
        close()
        eventGroup?.shutdownGracefully()
    }
}
